patches/server/1059-Implement-configurable-per-world-void-damage-thresho.patch

```diff
@@ -0,0 +1,103 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Axionize <154778082+Axionize@users.noreply.github.com>
+Date: Wed, 25 Sep 2024 04:42:19 -0400
+Subject: [PATCH] Implement configurable per-world void damage threshold height
+ & amount
+
+
+diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+index 790e3167cfa2511ef2fc707d9b9aa681a78d8ba0..47a0f9c64399d4eb8754851bde61731d36319e89 100644
+--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
++++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+@@ -409,6 +409,17 @@ public class WorldConfiguration extends ConfigurationPart {
+         public boolean disableExplosionKnockback = false;
+         public boolean generateFlatBedrock = false;
+         public FrostedIce frostedIce;
++        public VoidDamage voidDamage;
++
++        public class VoidDamage extends ConfigurationPart {
++            public boolean enabled = true;
++            public Damage damage;
++
++            public class Damage extends ConfigurationPart {
++                public double heightThreshold = -128.0;
++                public float amount = 4.0F;
++            }
++        }
+ 
+         public class FrostedIce extends ConfigurationPart {
+             public boolean enabled = true;
+diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
+index bc45c74987974b4828201e06fc8b1f3fbc0af8b4..305b1d80559077333cc5a31a7b217cf2f764cf5d 100644
+--- a/src/main/java/net/minecraft/world/entity/Entity.java
++++ b/src/main/java/net/minecraft/world/entity/Entity.java
+@@ -310,7 +310,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
+     protected int boardingCooldown;
+     @Nullable
+     private Entity vehicle;
+-    private Level level;
++    Level level;
+     public double xo;
+     public double yo;
+     public double zo;
+@@ -931,8 +931,12 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
+     }
+ 
+     public void checkBelowWorld() {
++        if (!this.level.paperConfig().environment.voidDamage.enabled) {
++            return;
++        }
++
+         // Paper start - Configurable nether ceiling damage
+-        if (this.getY() < (double) (this.level.getMinBuildHeight() - 64) || (this.level.getWorld().getEnvironment() == org.bukkit.World.Environment.NETHER
++        if (this.getY() < this.level.paperConfig().environment.voidDamage.damage.heightThreshold  || (this.level.getWorld().getEnvironment() == org.bukkit.World.Environment.NETHER
```

Comment:
In general, this breaks compatibility with worlds that do not conform to the expected "size" used when you calculated the default of -128.

This could be fixed by simply using DoubleOr.Default, allowing the default to evaluate to the specific min - 64 value.

Reply:
I might prefer just allowing the *offset* from the min build height to be configured. So the default would then be -64. Changing it to 0 would then make damage happen right at the min build height.